File: output/GPT4o-Mini/xCodeEval_4000/C++/Python/53e55dac69ffc2ce04d6eb21ec43d89b.py
import sys
from collections import defaultdict

sys.setrecursionlimit(10**6)

INF = 1011111111
LLINF = 1000111000111000111
EPS = 1e-10
mod = 1000000007
PI = 3.14159265358979323

g = defaultdict(set)
ct = 0
c = []
sz = [0] * 202020
used = [0] * 202020
n = 0
ans = []

def go(cur, iter=1):
    global ct
    sz[cur] = 1
    used[cur] = iter
    t = 0
    for to in g[cur]:
        if used[to] != iter:
            go(to, iter)
            sz[cur] += sz[to]
            t = max(t, sz[to])
    t = max(t, n - sz[cur])
    if t * 2 <= n:
        ct = cur
        c.append(cur)

st = []

def dfs2(cur, p=-1):
    used[cur] = True
    for to in g[cur]:
        if not used[to]:
            dfs2(to, cur)
    st.append((cur, p))

def solve(cur):
    global ct
    st.clear()
    dfs2(cur)
    st.pop()
    last = cur
    for v, par in st:
        if par != cur:
            ans.append((ct, last, v))
            ans.append((v, par, cur))
            last = v
    if last != cur:
        ans.append((ct, last, cur))

def main():
    global n
    n = int(input())
    for _ in range(n - 1):
        x, y = map(int, input().split())
        g[x].add(y)
        g[y].add(x)
    
    go(1)
    used[:] = [0] * len(used)
    assert len(c) > 0
    for x in c:
        used[x] = True
    for x in c:
        global ct
        ct = x
        for to in g[ct]:
            if not used[to]:
                solve(to)
    
    print(len(ans))
    assert len(ans) <= 2 * n
    for t in ans:
        print(t[0], t[1], t[2])

if __name__ == "__main__":
    main()

File: output/GPT4o-Mini/xCodeEval_4000/C++/C/53e55dac69ffc2ce04d6eb21ec43d89b.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define MAXN 202020
#define INF 1011111111

typedef struct {
    int a, b, c;
} tri;

typedef struct {
    int first, second;
} pii;

int g[MAXN][MAXN];
int g_size[MAXN];
int ct;
int sz[MAXN];
bool used[MAXN];
int n;
tri ans[MAXN * 2];
int ans_size = 0;

void go(int cur, int iter) {
    sz[cur] = 1;
    used[cur] = iter;
    int t = 0;
    for (int i = 0; i < g_size[cur]; i++) {
        int to = g[cur][i];
        if (used[to] != iter) {
            go(to, iter);
            sz[cur] += sz[to];
            t = (t > sz[to]) ? t : sz[to];
        }
    }
    t = (t > n - sz[cur]) ? t : (n - sz[cur]);
    if (t * 2 <= n) {
        ct = cur;
    }
}

pii st[MAXN];
int st_size = 0;

void dfs2(int cur, int p) {
    used[cur] = true;
    for (int i = 0; i < g_size[cur]; i++) {
        int to = g[cur][i];
        if (!used[to]) {
            dfs2(to, cur);
        }
    }
    st[st_size++] = (pii){cur, p};
}

void solve(int cur) {
    st_size = 0;
    dfs2(cur, -1);
    st_size--;
    int last = cur;
    for (int i = 0; i < st_size; i++) {
        int v = st[i].first;
        int par = st[i].second;
        if (par != cur) {
            ans[ans_size++] = (tri){ct, last, v};
            ans[ans_size++] = (tri){v, par, cur};
            last = v;
        }
    }
    if (last != cur) {
        ans[ans_size++] = (tri){ct, last, cur};
    }
}

int main() {
    scanf("%d", &n);
    for (int i = 0; i < n - 1; i++) {
        int x, y;
        scanf("%d %d", &x, &y);
        g[x][g_size[x]++] = y;
        g[y][g_size[y]++] = x;
    }
    go(1, 1);
    memset(used, 0, sizeof(used));
    for (int i = 0; i < n; i++) {
        if (g_size[i] > 0) {
            used[i] = true;
        }
    }
    for (int i = 0; i < n; i++) {
        if (used[i]) {
            ct = i;
            for (int j = 0; j < g_size[ct]; j++) {
                int to = g[ct][j];
                if (!used[to]) {
                    solve(to);
                }
            }
        }
    }
    printf("%d\n", ans_size);
    for (int i = 0; i < ans_size; i++) {
        printf("%d %d %d\n", ans[i].a, ans[i].b, ans[i].c);
    }
    return 0;
}
